src/ui/simulator/toolbox/components/datagrid/renderer/scenario-builder-hydro-final-levels-renderer.cpp

```diff
@@ -0,0 +1,86 @@
+/*
+** Copyright 2007-2018 RTE
+** Authors: Antares_Simulator Team
+**
+** This file is part of Antares_Simulator.
+**
+** Antares_Simulator is free software: you can redistribute it and/or modify
+** it under the terms of the GNU General Public License as published by
+** the Free Software Foundation, either version 3 of the License, or
+** (at your option) any later version.
+**
+** There are special exceptions to the terms and conditions of the
+** license as they are applied to this software. View the full text of
+** the exceptions in file COPYING.txt in the directory of this software
+** distribution
+**
+** Antares_Simulator is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with Antares_Simulator. If not, see <http://www.gnu.org/licenses/>.
+**
+** SPDX-License-Identifier: licenceRef-GPL3_WITH_RTE-Exceptions
+*/
+
+#include "scenario-builder-hydro-final-levels-renderer.h"
+#include "antares/study/scenario-builder/scBuilderUtils.h"
+
+using namespace Antares::Data::ScenarioBuilder;
```

Comment:
As the historic GUI will some day be removed, I guess we should be too strict with the code quality.
Anyway, here there must be a way to avoid code duplication with class **hydroLevelsScBuilderRenderer**

Reply:
Correction. In addition to:
hydroLevelsScBuilderRenderer::cellValue(int x, int y)
both other methods:
hydroLevelsScBuilderRenderer::cellValue(int x, int y, const Yuni::String& value)
hydroLevelsScBuilderRenderer::cellNumericValue(int x, int y)
have hardcoded pRules->hydroInitialLevels into them. I thought of adding this difference as a method parameter. But.
All these methods are inherited way back from abstract class IRenderer and are all pure virtual methods.
IRenderer is inherited in a huge number of other classes so if I poke/change the number of parameters for these functions here I'll have to do it everywhere else too.

Reply:
The UI is deprecated. Technical debt in the UI is fine since it will be discarded at some point.